astropy/io/ascii/tests/test_cds.py

```diff
@@ -0,0 +1,180 @@
+
+    out = StringIO()
+    t.write(out, format='ascii.cds')
+    # get the last section of table which will be the data.
+    lines = out.getvalue().splitlines()
+    i_secs = [i for i, s in enumerate(lines)
+              if s.startswith(('------', '======='))]
+    lines = lines[i_secs[-1]+1:]
+    assert lines == expOutput
+
+
+def test_write_ByteByByte_units():
+    t = ascii.read(test_dat)
+    colUnits = [None, u.C, u.kg, None, u.year]
+    t._set_column_attribute('unit', colUnits)
+    # add a column with magnitude units.
+    # note that magnitude has to be assigned for each value explicitly.
+    t['magnitude'] = [u.Magnitude(25), u.Magnitude(-9)]
+    colUnits.append(u.mag)
+    out = StringIO()
+    t.write(out, format='ascii.cds')
+    # read written table.
+    tRead = ascii.read(out.getvalue(), format='cds')
+    assert [tRead[col].unit for col in tRead.columns] == colUnits
+
+
+def test_write_readme_with_default_options():
+    expOutput = '''\
+                                                      (1st author ?, Date ?)
+================================================================================
+Title ?
+    Authors ?
+    ref ?
+================================================================================
+Keywords: 
+
+Objects:
+    -----------------------------------------
+       RA   (2000)   DE    Designation(s)
+    -----------------------------------------
+
+Abstract:
+  Abstract ?
+
+File Summary:
+--------------------------------------------------------------------------------
+ FileName    Lrecl   Records    Explanations
+--------------------------------------------------------------------------------
+ReadMe         80        . this file
+table          35        2          
+
+
+--------------------------------------------------------------------------------
+Byte-by-byte Description of file: table.dat
+--------------------------------------------------------------------------------
+ Bytes Format Units  Label     Explanations
+--------------------------------------------------------------------------------
+ 1- 8   A8     ---    names    Description of names             
+10-14   E5.1   ---    e       [-3160000.0/0.01] Description of e
+16-23   F8.5   ---    d       [22.25/27.25] Description of d    
+25-31   E7.1   ---    s       [-9e+34/2.0] Description of s     
+33-35   I3     ---    i       [-30/67] Description of i         
+
+--------------------------------------------------------------------------------
+
+See also:
+
+
+Acknowledgements:
+
+References:
+================================================================================
+     (prepared by author  / pyreadme )
+--------------------------------------------------------------------------------
+HD81809  1e-07  22.25608   2e+00  67
+HD103095 -3e+06 27.25000  -9e+34 -30
+'''
+    t = ascii.read(test_dat)
+    out = StringIO()
+    t.write(out, format='ascii.cds')
+    assert out.getvalue() == expOutput
+
+
+def test_write_empty_table():
```

Comment:
Also check this particular test. I ran it with other format writers and they output an empty table (with an empty header) for such a case. Ideally, the CDS writer should also do the same.
Right now, because this table has no columns, an ValueError gets raised when the total line width is calculated during writing of the ByteByByte description. I think there should be some mechanism to output an empty ByteByByte/ReadMe for an empty table.

Reply:
What do you think should happen here? Do we need to be able to write empty tables? It seems to be that that is a corner case we can deal with later if ever. How about `if len(table) == 0: raise NotImplementedError`?

Then, in the very end of this project, there can be a separate PR with the title "Enable writing empty CDS/MRT tables". Because it's not a bad thing to be able to write those, but we also have a lot of other things that are much higher on the priority list (e.g. mix-in columns: coordinates, time).

Reply:
> What do you think should happen here? Do we need to be able to write empty tables? It seems to be that that is a corner case we can deal with later if ever. How about `if len(table) == 0: raise NotImplementedError`?
Have put this in place. I would like to output unfilled ReadMe with no data rows in such cases. But I agree we can deal with empty table writing later.